File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/PropertyRegistry.swift
// PropertyRegistry.swift
// iOS StyleEngine — Phase 0 scaffolding.
//
// Purpose: a dispatch shell that maps `IRProperty.type` strings to dedicated
// extractor functions once they are migrated out of the monolithic
// `StyleBuilder.build(from:)` path.
//
// In Phase 0 this registry was intentionally empty: every property flowed
// through the legacy `StyleBuilder`. Later phases filled `migrated`, which the
// renderer consults DIRECTLY (`PropertyRegistry.migrated.contains`, see
// StyleBuilder.swift) — the `isLegacy(_:)` / `migratedCount` wrappers that
// once stood here had no caller and were deleted (retro P2a / round-2 F2,
// finding A6#8; the note at the end of this file records why).
//
// See `CLAUDE.md` → *Per-property contract* for the migration rules.

// Foundation gives us `Set<String>`, the only type we need at this stage.
import Foundation

/// Maps `IRProperty.type` → a typed piece of the `ComponentStyle` output.
///
/// Phase 0 scaffold: all properties still flow through the legacy
/// `StyleBuilder.build(from:)` monolith; this registry exists so future
/// phases can migrate properties one at a time without forking the
/// dispatch path.
///
/// See `CLAUDE.md` → *Per-property contract* for the migration contract.
enum PropertyRegistry {

    // MARK: - Migration ledger

    /// Property-type names that have been migrated out of `StyleBuilder`
    /// and into dedicated `{Property}Extractor.swift` files under
    /// `StyleEngine/{category}/`. Empty in Phase 0; filled by later phases.
    ///
    /// When a property is added here, the renderer routes its IR through
    /// the corresponding extractor instead of the legacy `StyleBuilder` path
    /// (StyleBuilder.swift tests membership in this set directly).
    static let migrated: Set<String> = Set<String>([
        // Phase 2 — spacing family. Padding/Margin physical+logical, Gap
        // longhands, MarginTrim. See runtimes/swiftui/.../StyleEngine/spacing/.
        "PaddingTop", "PaddingRight", "PaddingBottom", "PaddingLeft",
        "PaddingBlockStart", "PaddingBlockEnd",
        "PaddingInlineStart", "PaddingInlineEnd",
        "MarginTop", "MarginRight", "MarginBottom", "MarginLeft",
        "MarginBlockStart", "MarginBlockEnd",
        "MarginInlineStart", "MarginInlineEnd",
        "Gap", "RowGap", "ColumnGap",
        "MarginTrim",
        // Phase 3 — sizing family. Physical + logical sizing + aspect-ratio.
        // See runtimes/swiftui/.../StyleEngine/sizing/.
        "Width", "Height",
        "MinWidth", "MaxWidth", "MinHeight", "MaxHeight",
        "AspectRatio",
        "BlockSize", "InlineSize",
        "MinBlockSize", "MaxBlockSize",
        "MinInlineSize", "MaxInlineSize",
        // Phase 4 — colour + background + blend + isolation families.
        // See runtimes/swiftui/.../StyleEngine/{color,background,effects/blend,
        // performance}/.
        "BackgroundColor", "Color", "Opacity", "AccentColor", "CaretColor",
        // CSS Color HDR 1 §3.1 dynamic-range-limit. iOS: SwiftUI does
        // not expose HDR/SDR clamping at the View level so we accept it
        // for IR-coverage purposes only — the applier is identity until
        // CALayer.wantsExtendedDynamicRangeContent gating lands.
        "DynamicRangeLimit",
        "BackgroundImage",
        "BackgroundSize",
        "BackgroundPosition", "BackgroundPositionX", "BackgroundPositionY",
        // Logical-axis variants — folded onto Y/X in LTR horizontal mode
        // by BackgroundPositionExtractor.
        "BackgroundPositionBlock", "BackgroundPositionInline",
        "BackgroundRepeat",
        "BackgroundClip",
        "BackgroundOrigin",
        "BackgroundAttachment",
        "MixBlendMode", "BackgroundBlendMode",
        "Isolation",
        // Phase 5 — border family. Sides (physical + logical widths, colours,
        // styles + the three shorthand longhands), all eight radius corners,
        // outline quartet, border-image quintet, BoxShadow, and the keyword-
        // only miscellanies (BoxDecorationBreak, CornerShape, BorderBoundary).
        // See runtimes/swiftui/.../StyleEngine/borders/ and effects/shadow/.
        // Sides — shorthand longhands + 4 physical × 3 + 4 logical × 3 = 27.
        // A6#9: `BorderColor` is NOT listed — no `BorderColorProperty.kt`
        // exists in the IR catalogue (BorderColorExpander expands the
        // `border-color` shorthand to the four side longhands before the
        // longhand parser runs), so the name can never reach the wire.
        // BorderWidth/BorderStyle DO have IR classes and stay.
        "BorderWidth", "BorderStyle",
        "BorderTopWidth", "BorderTopColor", "BorderTopStyle",
        "BorderRightWidth", "BorderRightColor", "BorderRightStyle",
        "BorderBottomWidth", "BorderBottomColor", "BorderBottomStyle",
        "BorderLeftWidth", "BorderLeftColor", "BorderLeftStyle",
        "BorderBlockStartWidth", "BorderBlockStartColor", "BorderBlockStartStyle",
        "BorderBlockEndWidth",   "BorderBlockEndColor",   "BorderBlockEndStyle",
        "BorderInlineStartWidth", "BorderInlineStartColor", "BorderInlineStartStyle",
        "BorderInlineEndWidth",   "BorderInlineEndColor",   "BorderInlineEndStyle",
        // Radius — 4 physical + 4 logical = 8.
        "BorderTopLeftRadius", "BorderTopRightRadius",
        "BorderBottomRightRadius", "BorderBottomLeftRadius",
        "BorderStartStartRadius", "BorderStartEndRadius",
        "BorderEndEndRadius",   "BorderEndStartRadius",
        // Outline — 4.
        "OutlineWidth", "OutlineStyle", "OutlineColor", "OutlineOffset",
        // Border-image — 5.
        "BorderImageSource", "BorderImageSlice", "BorderImageWidth",
        "BorderImageOutset", "BorderImageRepeat",
        // BoxShadow — 1 (lives under effects/shadow/).
        "BoxShadow",
        // Miscellaneous keyword-only — 3.
        "BoxDecorationBreak", "CornerShape", "BorderBoundary",
        // Phase 6 — typography. Every rendering-capable triplet under
        // StyleEngine/typography/{font,font-variant,line,spacing,
        // decoration,wrapping,writing,other} is listed explicitly; the
        // five "unsupported" grouped extractors contribute their
        // {Group}Property.names lists via Set-union at the bottom.
        //
        // font/ (9)
        "FontSize", "FontFamily", "FontWeight", "FontStyle", "FontStretch",
        "FontFeatureSettings", "FontVariationSettings",
        "FontKerning", "FontOpticalSizing",
        // font-variant/ (7)
        "FontVariantCaps", "FontVariantNumeric", "FontVariantLigatures",
        "FontVariantEastAsian", "FontVariantPosition", "FontVariantAlternates",
        "FontVariantEmoji",
        // line/ (3)
        "LineHeight", "LineClamp", "MaxLines",
        // spacing/ (4)
        "LetterSpacing", "WordSpacing", "TabSize", "TextIndent",
        // decoration/ (8)
        "TextDecorationLine", "TextDecorationStyle", "TextDecorationColor",
        "TextDecorationThickness", "TextUnderlineOffset", "TextUnderlinePosition",
        "TextShadow", "TextTransform",
        // wrapping/ (11)
        "TextAlign", "TextAlignLast", "TextJustify", "TextWrap",
        "WhiteSpace", "WordBreak", "OverflowWrap", "LineBreak",
        "Hyphens", "HyphenateCharacter", "TextOverflow",
        // writing/ (5)
        "Direction", "UnicodeBidi", "WritingMode", "TextOrientation", "VerticalAlign",
        // other/ (2)
        "Quotes", "TextRendering",
    ])
    .union(UnsupportedSvgTypographyProperty.set)
    .union(UnsupportedPrintTypographyProperty.set)
    .union(UnsupportedRubyEmphasisProperty.set)
    .union(UnsupportedFontMetaProperty.set)
    .union(UnsupportedSpacingProperty.set)
    // Phase 7 — layout family (flexbox + grid + position + advanced +
    // root). The 60 names below are registered so the renderer ledger
    // reflects ownership; `LayoutExtractor` folds them into one
    // `LayoutAggregate` that ComponentRenderer consumes directly at
    // container-construction time. Retro P2b (A6#10/A6#3): the
    // "LayoutApplier is identity" half of this note named a scaffold enum
    // with no caller — deleted, so the aggregate's only consumer is now
    // the renderer. See StyleEngine/layout/LayoutExtractor.swift.
    .union(LayoutProperty.set)
    // Phase 8 — transforms, clip, visibility/overflow, filter, mask.
    // Each enum centralises its owned property-type names so the
    // extractors and self-tests stay in lockstep with this registry.
    .union(TransformsProperty.set)
    .union(ClipProperty.set)
    .union(VisibilityProperty.set)
    .union(FilterProperty.set)
    .union(MaskProperty.set)
    // Phase 9 — animations + transitions + view-timeline + view-transition
    // + timeline-scope (22 owned names) and the 3 scroll-timeline longhands.
    // Extractors live at StyleEngine/animations/ and StyleEngine/scrolling/.
    // Execution is NOT a view modifier: keyframes/transitions run as a
    // property-space pass (AnimationResolver + KeyframeInterpolator +
    // TransitionResolver) ahead of StyleBuilder — see AnimationsConfig.swift.
    // Retro P2b (A6#10): the identity `AnimationsApplier` /
    // `ScrollTimelineApplier` modifiers this note pointed at had no call
    // site (their `.engineAnimations` / `.engineScrollTimeline` chain
    // helpers were never attached) and were deleted.
    .union(AnimationsProperty.set)
    .union(ScrollTimelineProperty.set)
    // Phase 10 — long-tail sweep (~22 categories, ~150 property names).
    // Each category owns a grouped Config/Extractor/Applier triplet under
    // StyleEngine/<category>/. Nearly all appliers are identity-with-TODO
    // because SwiftUI has no analog; a small SwiftUI-capable subset
    // (scroll-snap, overscroll-behavior, pointer-events:none, user-
    // select:none) is flagged in the per-category applier headers for a
    // future wiring pass. The per-category Phase-10 fixtures are
    // `fixtures/properties/<category>/longtail.json`; the index this line
    // used to cite (`examples/properties/README-phase10.md`) was deleted by
    // the 2026-07-08 restructure, commit 02e4c457 (retro P2e).
    .union(ScrollingProperty.set)
    .union(SvgProperty.set)
    .union(UnsupportedSpeechProperty.set)
    .union(RenderingProperty.set)
    .union(UnsupportedPrintProperty.set)
    .union(UnsupportedPagingProperty.set)
    .union(UnsupportedRegionsProperty.set)
    .union(InteractionsProperty.set)
    .union(PerformanceProperty.set)
    .union(ColumnsProperty.set)
    // Wave 24 (lane GAPS-I) — css-gaps-1 gap decorations, claimed in the
    // SAME columns/ category as their column-rule-* twins. Only the NEW
    // names are unioned here: ColumnRuleColor/Style/Width are already in
    // ColumnsProperty.set above, and GapDecorationsExtractor reads them
    // without re-claiming them.
    .union(GapDecorationsProperty.set)
    .union(TableProperty.set)
    .union(ShapesProperty.set)
    .union(RhythmProperty.set)
    .union(UnsupportedNavigationProperty.set)
    .union(ImagesProperty.set)
    .union(AppearanceProperty.set)
    .union(CountersProperty.set)
    .union(ListsProperty.set)
    .union(ContainerProperty.set)
    .union(UnsupportedMathProperty.set)
    .union(ExperimentalProperty.set)
    .union(ContentProperty.set)
    .union(GlobalProperty.set)

    // MARK: - Query helpers

    // (Retro sweep P2a, finding A6#8) `isLegacy(_:)` and `migratedCount`
    // stood here and had ZERO references in Sources/, Tests/ and
    // apps/ios-harness. Their doc comments described consumers that do not
    // exist: the renderer decides dispatch with `PropertyRegistry.migrated
    // .contains(...)` directly (StyleBuilder.swift), there is no "rollout
    // dashboard", and the coverage report (tools/visual/coverage-audit.mjs)
    // scrapes FILES, never a runtime count. Both are deleted; the `migrated`
    // set they wrapped is untouched and still the live dispatch input.
    // The web twin of `migratedCount` was removed for the same reason
    // (A6#13, runtimes/web/src/engine/PropertyRegistry.ts).
}

File: runtimes/web/src/engine/PropertyRegistry.ts
/**
 * Maps IRProperty.type → a typed piece of the resulting CSSStyles.
 *
 * Phase 0 scaffold: all properties still flow through the legacy
 * `StyleBuilder.ts` monolith; this registry exists so future phases
 * can migrate properties one at a time without forking the dispatch
 * path.
 *
 * See `CLAUDE.md` → *Per-property contract* for the migration contract.
 */

/**
 * Property-type names that have been migrated out of `StyleBuilder.ts`
 * and into dedicated `{Property}Extractor.ts` files under
 * `engine/{category}/`. Empty in Phase 0; filled by later phases.
 */
export const migratedProperties = new Set<string>([
  // Spacing — Phase 2 migration. See engine/spacing/*.
  'PaddingTop', 'PaddingRight', 'PaddingBottom', 'PaddingLeft',
  'PaddingBlockStart', 'PaddingBlockEnd', 'PaddingInlineStart', 'PaddingInlineEnd',
  'MarginTop', 'MarginRight', 'MarginBottom', 'MarginLeft',
  'MarginBlockStart', 'MarginBlockEnd', 'MarginInlineStart', 'MarginInlineEnd',
  'Gap', 'RowGap', 'ColumnGap',
  'MarginTrim',
  // Sizing — Phase 3 migration. See engine/sizing/*.
  'Width', 'Height', 'MinWidth', 'MaxWidth', 'MinHeight', 'MaxHeight',
  'AspectRatio',
  'BlockSize', 'InlineSize',
  'MinBlockSize', 'MaxBlockSize', 'MinInlineSize', 'MaxInlineSize',
  // css-sizing-3 §4 — real triplet since issue #38 (engine/sizing/BoxSizing*).
  'BoxSizing',
  // Colors + background — Phase 4 migration. See engine/color/*, engine/background/*,
  // engine/effects/blend/*, engine/performance/*.
  'BackgroundColor', 'Color', 'Opacity', 'AccentColor', 'CaretColor',
  // CSS Color HDR 1 §3.1 — real triplet since issue #38
  // (engine/color/DynamicRangeLimit*): csstype-widened one-key emitter.
  'DynamicRangeLimit',
  // css-color-adjust-1 §2 — real triplet since issue #38
  // (engine/color/ColorScheme*): emits the `color-scheme` list verbatim.
  'ColorScheme',
  'BackgroundImage', 'BackgroundSize',
  'BackgroundPosition', 'BackgroundPositionX', 'BackgroundPositionY',
  // Logical-axis variants — extractor folds them onto the physical axes.
  'BackgroundPositionBlock', 'BackgroundPositionInline',
  'BackgroundRepeat', 'BackgroundClip', 'BackgroundOrigin', 'BackgroundAttachment',
  'MixBlendMode', 'BackgroundBlendMode', 'Isolation',
  // Borders — Phase 5 migration. See engine/borders/* and engine/effects/shadow/BoxShadow*.
  // Sides (24): width/style/color for 4 physical + 4 logical edges.
  'BorderTopWidth', 'BorderRightWidth', 'BorderBottomWidth', 'BorderLeftWidth',
  'BorderBlockStartWidth', 'BorderBlockEndWidth', 'BorderInlineStartWidth', 'BorderInlineEndWidth',
  'BorderTopStyle', 'BorderRightStyle', 'BorderBottomStyle', 'BorderLeftStyle',
  'BorderBlockStartStyle', 'BorderBlockEndStyle', 'BorderInlineStartStyle', 'BorderInlineEndStyle',
  'BorderTopColor', 'BorderRightColor', 'BorderBottomColor', 'BorderLeftColor',
  'BorderBlockStartColor', 'BorderBlockEndColor', 'BorderInlineStartColor', 'BorderInlineEndColor',
  // Radius (8): 4 physical corners + 4 logical corners.
  'BorderTopLeftRadius', 'BorderTopRightRadius', 'BorderBottomRightRadius', 'BorderBottomLeftRadius',
  'BorderStartStartRadius', 'BorderStartEndRadius', 'BorderEndStartRadius', 'BorderEndEndRadius',
  // Image (5): Source / Slice / Width / Outset / Repeat.
  'BorderImageSource', 'BorderImageSlice', 'BorderImageWidth', 'BorderImageOutset', 'BorderImageRepeat',
  // Outline (4): Width / Style / Color / Offset.
  'OutlineWidth', 'OutlineStyle', 'OutlineColor', 'OutlineOffset',
  // Shadow (1): BoxShadow mirrored under engine/effects/shadow/.
  'BoxShadow',
  // Misc (3): BoxDecorationBreak / CornerShape / BorderBoundary.
  'BoxDecorationBreak', 'CornerShape', 'BorderBoundary',
  // Typography — Phase 6 migration.  See engine/typography/*.  109 properties
  // (CaretColor was already migrated under engine/color/* in Phase 4, so it's
  // not re-added here).
  'AlignmentBaseline', 'BaselineShift', 'BaselineSource', 'BlockEllipsis',
  'Direction', 'DominantBaseline', 'DominantBaselineAdjust',
  'FontDisplay', 'FontFamily', 'FontFeatureSettings', 'FontKerning',
  'FontLanguageOverride', 'FontMaxSize', 'FontMinSize', 'FontNamedInstance',
  'FontOpticalSizing', 'FontPalette', 'FontSize', 'FontSizeAdjust',
  'FontSmooth', 'FontStretch', 'FontStyle',
  'FontSynthesisPosition', 'FontSynthesisSmallCaps', 'FontSynthesisStyle', 'FontSynthesisWeight',
  'FontVariantAlternates', 'FontVariantCaps', 'FontVariantEastAsian',
  'FontVariantEmoji', 'FontVariantLigatures', 'FontVariantNumeric', 'FontVariantPosition',
  'FontVariationSettings', 'FontWeight',
  'GlyphOrientationHorizontal', 'GlyphOrientationVertical', 'HangingPunctuation',
  'HyphenateCharacter', 'HyphenateLimitChars', 'HyphenateLimitLast',
  'HyphenateLimitLines', 'HyphenateLimitZone', 'Hyphens',
  'InitialLetter', 'InitialLetterAlign', 'Kerning',
  'LetterSpacing', 'LineBreak', 'LineClamp', 'LineGrid',
  'LineHeight', 'LineHeightStep', 'LineSnap', 'MaxLines',
  'Orphans', 'OverflowWrap', 'Quotes',
  'RubyAlign', 'RubyMerge', 'RubyOverhang', 'RubyPosition',
  'TabSize', 'TextAlign', 'TextAlignAll', 'TextAlignLast', 'TextAnchor',
  'TextAutospace', 'TextBoxEdge', 'TextBoxTrim', 'TextCombineUpright',
  'TextDecorationColor', 'TextDecorationLine', 'TextDecorationSkip',
  'TextDecorationSkipInk', 'TextDecorationStyle', 'TextDecorationThickness',
  'TextEmphasis', 'TextEmphasisColor', 'TextEmphasisPosition', 'TextEmphasisStyle',
  'TextGroupAlign', 'TextIndent', 'TextJustify', 'TextOrientation',
  'TextOverflow', 'TextRendering', 'TextShadow', 'TextSizeAdjust',
  'TextSpaceCollapse', 'TextSpaceTrim', 'TextSpacing', 'TextSpacingTrim',
  'TextTransform', 'TextUnderlineOffset', 'TextUnderlinePosition',
  'TextWrap', 'TextWrapMode', 'TextWrapStyle',
  'UnicodeBidi', 'VerticalAlign', 'VerticalAlignLast',
  'WhiteSpace', 'WhiteSpaceCollapse', 'Widows',
  'WordBreak', 'WordSpaceTransform', 'WordSpacing', 'WordWrap', 'WritingMode',
  // Layout — Phase 7 migration.  See engine/layout/*.  55 properties covering
  // flexbox (11), grid (14), position/edges (10), anchor-positioning + motion-
  // path advanced (16), plus the four root flow/top-layer keywords
  // (Clear/Float/Overlay/ReadingFlow).  All route through
  // `applyLayoutPhase7` in engine/layout/_dispatch.ts.
  'Clear', 'Float', 'Overlay', 'ReadingFlow',
  'AnchorName', 'AnchorScope', 'InsetArea',
  'OffsetAnchor', 'OffsetDistance', 'OffsetPath', 'OffsetPosition', 'OffsetRotate',
  'PositionAnchor', 'PositionArea', 'PositionFallback',
  'PositionTry', 'PositionTryFallbacks', 'PositionTryOptions', 'PositionTryOrder',
  'PositionVisibility',
  'Display', 'FlexDirection', 'FlexWrap', 'FlexBasis', 'FlexGrow', 'FlexShrink',
  'JustifyContent', 'AlignItems', 'AlignContent', 'AlignSelf', 'Order',
  'GridTemplateColumns', 'GridTemplateRows', 'GridTemplateAreas',
  'GridAutoColumns', 'GridAutoRows', 'GridAutoFlow',
  'GridColumnStart', 'GridColumnEnd', 'GridRowStart', 'GridRowEnd',
  'JustifyItems', 'JustifySelf',
  'AlignTracks', 'JustifyTracks', 'MasonryAutoFlow',
  // grid-auto-track (css-grid-3 masonry draft; its longhand parser IS
  // registered, so the type reaches the wire) — real triplet since issue
  // #38 (engine/layout/grid/GridAutoTrack*), standalone from GridExtractor.
  'GridAutoTrack',
  // -webkit-box-orient (legacy, line-clamp companion) — real triplet since
  // issue #38 (engine/layout/flexbox/BoxOrient*). The old note claimed the
  // legacy path emitted it, but membership in this set disabled that path.
  'BoxOrient',
  'Position', 'Top', 'Right', 'Bottom', 'Left',
  'InsetBlockStart', 'InsetBlockEnd', 'InsetInlineStart', 'InsetInlineEnd',
  'ZIndex',
  // Effects + transforms — Phase 8 migration.  See engine/transforms/*,
  // engine/effects/{clip,filter,mask}/*, engine/visibility/*.  38 properties.
  // Transforms (10)
  'Transform', 'Rotate', 'Scale', 'Translate', 'TransformOrigin', 'TransformBox',
  'TransformStyle', 'Perspective', 'PerspectiveOrigin', 'BackfaceVisibility',
  // Clip + visibility + overflow (10)
  'ClipPath', 'ClipPathGeometryBox', 'ClipRule', 'Clip',
  'Visibility', 'Overflow', 'OverflowX', 'OverflowY', 'OverflowBlock', 'OverflowInline',
  // Filter (2)
  'Filter', 'BackdropFilter',
  // Mask (16)
  'MaskImage', 'MaskMode', 'MaskRepeat', 'MaskPosition', 'MaskPositionX', 'MaskPositionY',
  'MaskSize', 'MaskOrigin', 'MaskClip', 'MaskComposite', 'MaskType',
  'MaskBorderSource', 'MaskBorderSlice', 'MaskBorderWidth', 'MaskBorderOutset',
  'MaskBorderRepeat', 'MaskBorderMode',
  // Phase 9 — animations/transitions/view-timeline/view-transition (26) +
  // scroll-timeline (3).  See engine/animations/* and engine/scrolling/*.
  // Most pass through to native CSS; L2 keys are csstype-widened in the
  // applier (see `as unknown as CSSProperties` comments inline).
  'AnimationName', 'AnimationDuration', 'AnimationDelay', 'AnimationIterationCount',
  'AnimationDirection', 'AnimationFillMode', 'AnimationPlayState', 'AnimationComposition',
  'AnimationTimingFunction', 'AnimationTimeline', 'AnimationRange', 'AnimationRangeStart',
  'AnimationRangeEnd',
  'TransitionProperty', 'TransitionDuration', 'TransitionDelay', 'TransitionTimingFunction',
  'TransitionBehavior',
  'TimelineScope', 'ViewTimeline', 'ViewTimelineAxis', 'ViewTimelineInset',
  'ViewTimelineName', 'ViewTransitionName', 'ViewTransitionClass', 'ViewTransitionGroup',
  'ScrollTimeline', 'ScrollTimelineName', 'ScrollTimelineAxis',
  // Phase 10 — long tail (~170 properties across 22 categories).  All route
  // through `engine/<category>/_dispatch.ts` + the per-category
  // `apply<Category>Phase10(properties)` entry point.  Web is the privileged
  // platform: nearly all of these pass straight through to native CSS.  A
  // subset are csstype-widened because they're CSS L4 / spec-only / dropped-
  // from-spec; see the individual appliers for MDN links.
  // ── scrolling (44, non-timeline) ─────────────────────────────────────
  'ScrollBehavior', 'ScrollSnapType', 'ScrollSnapAlign', 'ScrollSnapStop',
  'ScrollPadding', 'ScrollPaddingTop', 'ScrollPaddingRight', 'ScrollPaddingBottom', 'ScrollPaddingLeft',
  'ScrollPaddingBlock', 'ScrollPaddingBlockStart', 'ScrollPaddingBlockEnd',
  'ScrollPaddingInline', 'ScrollPaddingInlineStart', 'ScrollPaddingInlineEnd',
  'ScrollMargin', 'ScrollMarginTop', 'ScrollMarginRight', 'ScrollMarginBottom', 'ScrollMarginLeft',
  'ScrollMarginBlock', 'ScrollMarginBlockStart', 'ScrollMarginBlockEnd',
  'ScrollMarginInline', 'ScrollMarginInlineStart', 'ScrollMarginInlineEnd',
  'OverscrollBehavior', 'OverscrollBehaviorX', 'OverscrollBehaviorY',
  'OverscrollBehaviorBlock', 'OverscrollBehaviorInline',
  'ScrollbarWidth', 'ScrollbarColor', 'ScrollbarGutter',
  'OverflowAnchor', 'OverflowClipMargin',
  'ScrollStart', 'ScrollStartX', 'ScrollStartY', 'ScrollStartBlock', 'ScrollStartInline',
  'ScrollStartTarget', 'ScrollMarkerGroup', 'ScrollTargetGroup',
  // scroll-start-target axis variants — real triplets since issue #38
  // (engine/scrolling/ScrollStartTarget{Block,Inline,X,Y}*).
  'ScrollStartTargetBlock', 'ScrollStartTargetInline',
  'ScrollStartTargetX', 'ScrollStartTargetY',
  // ── svg (36) ─────────────────────────────────────────────────────────
  'Fill', 'FillOpacity', 'FillRule',
  'Stroke', 'StrokeOpacity', 'StrokeWidth', 'StrokeDasharray', 'StrokeDashoffset',
  'StrokeLinecap', 'StrokeLinejoin', 'StrokeMiterlimit',
  'StopColor', 'StopOpacity', 'FloodColor', 'FloodOpacity', 'LightingColor',
  'ColorInterpolation', 'ColorInterpolationFilters', 'ColorRendering',
  'MarkerStart', 'MarkerMid', 'MarkerEnd', 'Marker',
  'PaintOrder', 'ShapeRendering', 'VectorEffect', 'BufferedRendering', 'EnableBackground',
  'Cx', 'Cy', 'R', 'Rx', 'Ry', 'X', 'Y', 'D',
  // marker-side (SVG 2 draft) — real triplet since issue #38
  // (engine/svg/MarkerSide*): csstype-widened pass-through.
  'MarkerSide',
  // ── speech (30) ──────────────────────────────────────────────────────
  'Volume', 'Speak', 'SpeakAs', 'SpeakHeader', 'SpeakNumeral', 'SpeakPunctuation',
  'Pause', 'PauseBefore', 'PauseAfter', 'Rest', 'RestBefore', 'RestAfter',
  'Cue', 'CueBefore', 'CueAfter',
  'VoiceFamily', 'VoiceRate', 'VoicePitch', 'VoiceRange', 'VoiceStress',
  'VoiceVolume', 'VoiceDuration', 'VoiceBalance',
  'Pitch', 'PitchRange', 'Richness', 'Stress', 'SpeechRate', 'Azimuth', 'Elevation',
  // ── rendering (9) ────────────────────────────────────────────────────
  'ContentVisibility', 'FieldSizing', 'ForcedColorAdjust', 'PrintColorAdjust',
  'ImageOrientation', 'ImageResolution', 'InputSecurity', 'InterpolateSize', 'Zoom',
  // ── print (11) ───────────────────────────────────────────────────────
  'Bleed', 'BookmarkLabel', 'BookmarkLevel', 'BookmarkState', 'BookmarkTarget',
  'FootnoteDisplay', 'FootnotePolicy', 'Leader', 'Marks', 'Page', 'Size',
  // ── regions (10) ─────────────────────────────────────────────────────
  'FlowInto', 'FlowFrom', 'RegionFragment', 'Continue', 'CopyInto',
  'WrapFlow', 'WrapThrough', 'WrapBefore', 'WrapAfter', 'WrapInside',
  // ── interactions (8) ─────────────────────────────────────────────────
  'Cursor', 'PointerEvents', 'UserSelect', 'TouchAction', 'Resize',
  'Interactivity', 'Caret', 'CaretShape',
  // ── performance (7) — Isolation was Phase 4 ──────────────────────────
  'Contain', 'WillChange',
  'ContainIntrinsicSize', 'ContainIntrinsicWidth', 'ContainIntrinsicHeight',
  'ContainIntrinsicBlockSize', 'ContainIntrinsicInlineSize',
  // ── columns (7) — ColumnGap was Phase 2 ──────────────────────────────
  'ColumnCount', 'ColumnWidth',
  'ColumnRuleStyle', 'ColumnRuleWidth', 'ColumnRuleColor',
  'ColumnSpan', 'ColumnFill',
  // ── paging (7) ───────────────────────────────────────────────────────
  'BreakBefore', 'BreakAfter', 'BreakInside',
  'PageBreakBefore', 'PageBreakAfter', 'PageBreakInside', 'MarginBreak',
  // ── table (5) ────────────────────────────────────────────────────────
  'BorderCollapse', 'BorderSpacing', 'CaptionSide', 'EmptyCells', 'TableLayout',
  // ── shapes (5) ───────────────────────────────────────────────────────
  'ShapeOutside', 'ShapeMargin', 'ShapePadding', 'ShapeImageThreshold', 'ShapeInside',
  // ── rhythm (5) — block-step family, widen ────────────────────────────
  'BlockStep', 'BlockStepAlign', 'BlockStepInsert', 'BlockStepRound', 'BlockStepSize',
  // ── navigation (5) — dropped from spec, widen ────────────────────────
  'NavUp', 'NavDown', 'NavLeft', 'NavRight', 'ReadingOrder',
  // ── images (4) ───────────────────────────────────────────────────────
  'ImageRendering', 'ObjectFit', 'ObjectPosition', 'ObjectViewBox',
  // ── appearance (4) ───────────────────────────────────────────────────
  'Appearance', 'AppearanceVariant', 'ColorAdjust', 'ImageRenderingQuality',
  // ── counters (3) ─────────────────────────────────────────────────────
  'CounterIncrement', 'CounterReset', 'CounterSet',
  // ── lists (3) ────────────────────────────────────────────────────────
  'ListStyleType', 'ListStylePosition', 'ListStyleImage',
  // ── container (3) ────────────────────────────────────────────────────
  'Container', 'ContainerName', 'ContainerType',
  // ── math (3), widen ──────────────────────────────────────────────────
  'MathStyle', 'MathShift', 'MathDepth',
  // ── experimental (3), widen ──────────────────────────────────────────
  'PresentationLevel', 'Running', 'StringSet',
  // ── content (1) — Quotes already Phase 6 typography ──────────────────
  'Content',
  // ── global (1) ───────────────────────────────────────────────────────
  'All',
  // ── coverage-only claims — SHORTHAND types that can never reach the
  // wire: PropertiesParser expands shorthands via ShorthandRegistry
  // BEFORE the longhand parser registry runs (PropertiesParser.kt step
  // 2), and every one of these names has a registered Expander
  // (ShorthandRegistry.kt), so the converter only ever emits their
  // longhands — which are all individually claimed above. Verified in
  // the issue #38 audit; every OTHER former coverage-only entry
  // (BoxSizing, ColorScheme, DynamicRangeLimit, BoxOrient,
  // GridAutoTrack, ScrollStartTarget{Block,Inline,X,Y}, MarkerSide)
  // turned out to be reachable on the wire and now has a real triplet
  // in its category section.
  // ── borders (2) — expand to the 8 side longhands ─────────────────────
  'BorderWidth', 'BorderStyle',
  // ── layout (3) — Motion-Path 1 `offset` + grid shorthand bundles ─────
  'Offset', 'GridArea', 'GridTemplate',
]);

/**
 * Returns true when the given IR property type is still served by the
 * legacy `StyleBuilder`. Used by the renderer during transition.
 */
export function isLegacyProperty(propertyType: string): boolean {
  // Negate membership: anything not yet migrated stays on the legacy path.
  return !migratedProperties.has(propertyType);
}

/** Count of migrated properties — exposed for the coverage report. */
export const migratedCount: number = migratedProperties.size;
